File: python/src/virtuus/_python/gsi.py
"""Global Secondary Index implementation."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable, Optional


@dataclass(frozen=True)
class _GsiEntry:
    pk: str
    sort_value: Optional[Any]


class GSI:
    """Global Secondary Index with hash partition and optional range key.

    :param name: Name of the index.
    :type name: str
    :param partition_key: Field used as the partition key.
    :type partition_key: str
    :param sort_key: Optional field used as the sort (range) key.
    :type sort_key: str | None
    """

    def __init__(self, name: str, partition_key: str, sort_key: Optional[str] = None):
        self._name = name
        self._partition_key = partition_key
        self._sort_key = sort_key
        self._buckets: dict[Any, list[_GsiEntry]] = {}

    @property
    def name(self) -> str:
        """Return the index name.

        :return: Index name.
        :rtype: str
        """

        return self._name

    @property
    def partition_key(self) -> str:
        """Return the partition key field.

        :return: Partition key field name.
        :rtype: str
        """

        return self._partition_key

    @property
    def sort_key(self) -> Optional[str]:
        """Return the sort key field if present.

        :return: Sort key field name or None.
        :rtype: str | None
        """

        return self._sort_key

    def put(self, pk: str, record: dict[str, Any]) -> None:
        """Insert a record into the index.

        :param pk: Primary key of the record.
        :type pk: str
        :param record: Record data.
        :type record: dict[str, Any]
        :return: None
        :rtype: None
        """

        partition_value = record.get(self._partition_key)
        if partition_value is None:
            return
        sort_value = self._extract_sort_value(record)
        if self._sort_key is not None and sort_value is None:
            return
        bucket = self._buckets.setdefault(partition_value, [])
        bucket.append(_GsiEntry(pk=pk, sort_value=sort_value))

    def remove(self, pk: str, record: dict[str, Any]) -> None:
        """Remove a record from the index.

        :param pk: Primary key of the record.
        :type pk: str
        :param record: Record data.
        :type record: dict[str, Any]
        :return: None
        :rtype: None
        """

        partition_value = record.get(self._partition_key)
        if partition_value is None:
            return
        sort_value = self._extract_sort_value(record)
        if self._sort_key is not None and sort_value is None:
            return
        bucket = self._buckets.get(partition_value)
        if not bucket:
            return
        self._buckets[partition_value] = [
            entry
            for entry in bucket
            if not (entry.pk == pk and entry.sort_value == sort_value)
        ]
        if not self._buckets[partition_value]:
            self._buckets.pop(partition_value, None)

    def update(self, pk: str, old_record: dict[str, Any], new_record: dict[str, Any]) -> None:
        """Update an indexed record.

        :param pk: Primary key of the record.
        :type pk: str
        :param old_record: Previous record data.
        :type old_record: dict[str, Any]
        :param new_record: Updated record data.
        :type new_record: dict[str, Any]
        :return: None
        :rtype: None
        """

        self.remove(pk, old_record)
        self.put(pk, new_record)

    def query(
        self,
        partition_value: Any,
        sort_condition: Optional[Callable[[Any], bool]] = None,
        sort_direction: str = "asc",
    ) -> list[str]:
        """Query by partition key with optional sort condition and direction.

        :param partition_value: Partition key value to query.
        :type partition_value: Any
        :param sort_condition: Optional predicate over sort values.
        :type sort_condition: Callable[[Any], bool] | None
        :param sort_direction: "asc" or "desc".
        :type sort_direction: str
        :return: List of primary keys.
        :rtype: list[str]
        :raises ValueError: If sort_direction is not "asc" or "desc".
        """

        if sort_direction not in {"asc", "desc"}:
            raise ValueError("sort_direction must be 'asc' or 'desc'")
        bucket = list(self._buckets.get(partition_value, []))
        if sort_condition is not None:
            bucket = [
                entry
                for entry in bucket
                if entry.sort_value is not None and sort_condition(entry.sort_value)
            ]
        if self._sort_key is not None:
            bucket = sorted(bucket, key=lambda entry: _order_key(entry.sort_value))
        if sort_direction == "desc":
            bucket = list(reversed(bucket))
        return [entry.pk for entry in bucket]

    def _extract_sort_value(self, record: dict[str, Any]) -> Optional[Any]:
        if self._sort_key is None:
            return None
        return record.get(self._sort_key)


def _value_rank(value: Any) -> int:
    if value is None:
        return 0
    if isinstance(value, bool):
        return 1
    if isinstance(value, (int, float)):
        return 2
    if isinstance(value, str):
        return 3
    if isinstance(value, list):
        return 4
    if isinstance(value, dict):
        return 5
    return 6


def _normalize_for_order(value: Any) -> Any:
    if value is None:
        return ""
    if isinstance(value, bool):
        return 1 if value else 0
    if isinstance(value, (int, float, str)):
        return value
    if isinstance(value, list):
        return tuple(_order_key(item) for item in value)
    if isinstance(value, dict):
        return tuple(sorted((str(k), _order_key(v)) for k, v in value.items()))
    return str(value)


def _order_key(value: Any) -> tuple[int, Any]:
    return _value_rank(value), _normalize_for_order(value)

File: rust/src/gsi.rs
//! Global Secondary Index implementation.

use std::collections::HashMap;

use serde_json::Value;

use crate::sort::{OrderedValue, SortCondition};

#[derive(Debug, Clone, PartialEq)]
struct GsiEntry {
    pk: String,
    sort_value: Option<Value>,
}

/// Global Secondary Index with hash partition and optional range key.
#[derive(Debug, Clone)]
pub struct Gsi {
    name: String,
    partition_key: String,
    sort_key: Option<String>,
    buckets: HashMap<String, Vec<GsiEntry>>,
}

impl Gsi {
    /// Create a new GSI definition.
    ///
    /// :param name: Index name.
    /// :param partition_key: Partition key field.
    /// :param sort_key: Optional sort key field.
    pub fn new(name: &str, partition_key: &str, sort_key: Option<&str>) -> Self {
        Self {
            name: name.to_string(),
            partition_key: partition_key.to_string(),
            sort_key: sort_key.map(|s| s.to_string()),
            buckets: HashMap::new(),
        }
    }

    /// Return the index name.
    pub fn name(&self) -> &str {
        &self.name
    }

    /// Return the partition key field name.
    pub fn partition_key(&self) -> &str {
        &self.partition_key
    }

    /// Return the sort key field name, if present.
    pub fn sort_key(&self) -> Option<&str> {
        self.sort_key.as_deref()
    }

    /// Insert a record into the index.
    pub fn put(&mut self, pk: &str, record: &Value) {
        let partition_value = match get_field(record, &self.partition_key) {
            Some(value) => value,
            None => return,
        };
        let sort_value = self.extract_sort_value(record);
        if self.sort_key.is_some() && sort_value.is_none() {
            return;
        }
        let key = partition_key(&partition_value);
        let bucket = self.buckets.entry(key).or_default();
        bucket.push(GsiEntry {
            pk: pk.to_string(),
            sort_value,
        });
    }

    /// Remove a record from the index.
    pub fn remove(&mut self, pk: &str, record: &Value) {
        let partition_value = match get_field(record, &self.partition_key) {
            Some(value) => value,
            None => return,
        };
        let sort_value = self.extract_sort_value(record);
        if self.sort_key.is_some() && sort_value.is_none() {
            return;
        }
        let key = partition_key(&partition_value);
        let bucket = match self.buckets.get_mut(&key) {
            Some(bucket) => bucket,
            None => return,
        };
        bucket.retain(|entry| !(entry.pk == pk && entry.sort_value == sort_value));
        if bucket.is_empty() {
            self.buckets.remove(&key);
        }
    }

    /// Update an indexed record.
    pub fn update(&mut self, pk: &str, old_record: &Value, new_record: &Value) {
        self.remove(pk, old_record);
        self.put(pk, new_record);
    }

    /// Query by partition key with optional sort condition and direction.
    pub fn query(
        &self,
        partition_value: &Value,
        sort_condition: Option<&SortCondition>,
        descending: bool,
    ) -> Vec<String> {
        let key = partition_key(partition_value);
        let mut bucket = match self.buckets.get(&key) {
            Some(entries) => entries.clone(),
            None => return Vec::new(),
        };
        if let Some(condition) = sort_condition {
            bucket.retain(|entry| match entry.sort_value.as_ref() {
                Some(value) => condition.evaluate(value),
                None => false,
            });
        }
        if self.sort_key.is_some() {
            bucket.sort_by(|a, b| {
                OrderedValue(a.sort_value.clone().unwrap_or(Value::Null))
                    .cmp(&OrderedValue(b.sort_value.clone().unwrap_or(Value::Null)))
            });
        }
        if descending {
            bucket.reverse();
        }
        bucket.into_iter().map(|entry| entry.pk).collect()
    }

    fn extract_sort_value(&self, record: &Value) -> Option<Value> {
        let sort_key = self.sort_key.as_ref()?;
        get_field(record, sort_key)
    }
}

fn get_field(record: &Value, key: &str) -> Option<Value> {
    match record {
        Value::Object(map) => map.get(key).cloned(),
        _ => None,
    }
}

fn partition_key(value: &Value) -> String {
    serde_json::to_string(value).unwrap_or_else(|_| format!("\"{}\"", value))
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn create_hash_only_gsi() {
        let gsi = Gsi::new("by_status", "status", None);
        assert_eq!(gsi.name(), "by_status");
        assert_eq!(gsi.partition_key(), "status");
        assert!(gsi.sort_key().is_none());
    }

    #[test]
    fn create_hash_range_gsi() {
        let gsi = Gsi::new("by_org", "org_id", Some("created_at"));
        assert_eq!(gsi.sort_key(), Some("created_at"));
    }

    #[test]
    fn put_and_query_hash_only() {
        let mut gsi = Gsi::new("by_status", "status", None);
        gsi.put("user-1", &json!({"status": "active"}));
        let result = gsi.query(&json!("active"), None, false);
        assert_eq!(result, vec!["user-1".to_string()]);
    }

    #[test]
    fn missing_partition_key_is_skipped() {
        let mut gsi = Gsi::new("by_status", "status", None);
        gsi.put("user-1", &json!({"name": "Alice"}));
        let result = gsi.query(&json!("active"), None, false);
        assert!(result.is_empty());
    }

    #[test]
    fn missing_sort_key_is_skipped() {
        let mut gsi = Gsi::new("by_org", "org_id", Some("created_at"));
        gsi.put("user-1", &json!({"org_id": "org-a"}));
        let result = gsi.query(&json!("org-a"), None, false);
        assert!(result.is_empty());
    }

    #[test]
    fn remove_entry() {
        let mut gsi = Gsi::new("by_status", "status", None);
        gsi.put("user-1", &json!({"status": "active"}));
        gsi.remove("user-1", &json!({"status": "active"}));
        let result = gsi.query(&json!("active"), None, false);
        assert!(result.is_empty());
    }

    #[test]
    fn update_reindexes_partition() {
        let mut gsi = Gsi::new("by_status", "status", None);
        gsi.put("user-1", &json!({"status": "active"}));
        gsi.update(
            "user-1",
            &json!({"status": "active"}),
            &json!({"status": "inactive"}),
        );
        let active = gsi.query(&json!("active"), None, false);
        let inactive = gsi.query(&json!("inactive"), None, false);
        assert!(active.is_empty());
        assert_eq!(inactive, vec!["user-1".to_string()]);
    }

    #[test]
    fn hash_range_sorted_query() {
        let mut gsi = Gsi::new("by_org", "org_id", Some("created_at"));
        gsi.put(
            "user-1",
            &json!({"org_id": "org-a", "created_at": "2025-03-01"}),
        );
        gsi.put(
            "user-2",
            &json!({"org_id": "org-a", "created_at": "2025-01-15"}),
        );
        gsi.put(
            "user-3",
            &json!({"org_id": "org-a", "created_at": "2025-06-20"}),
        );
        let result = gsi.query(&json!("org-a"), None, false);
        assert_eq!(
            result,
            vec![
                "user-2".to_string(),
                "user-1".to_string(),
                "user-3".to_string()
            ]
        );
    }

    #[test]
    fn hash_range_descending_query() {
        let mut gsi = Gsi::new("by_org", "org_id", Some("created_at"));
        gsi.put(
            "user-1",
            &json!({"org_id": "org-a", "created_at": "2025-01-01"}),
        );
        gsi.put(
            "user-2",
            &json!({"org_id": "org-a", "created_at": "2025-06-01"}),
        );
        gsi.put(
            "user-3",
            &json!({"org_id": "org-a", "created_at": "2025-12-01"}),
        );
        let result = gsi.query(&json!("org-a"), None, true);
        assert_eq!(
            result,
            vec![
                "user-3".to_string(),
                "user-2".to_string(),
                "user-1".to_string()
            ]
        );
    }

    #[test]
    fn hash_range_sort_condition() {
        let mut gsi = Gsi::new("by_org", "org_id", Some("created_at"));
        gsi.put(
            "user-1",
            &json!({"org_id": "org-a", "created_at": "2025-01-01"}),
        );
        gsi.put(
            "user-2",
            &json!({"org_id": "org-a", "created_at": "2025-06-01"}),
        );
        gsi.put(
            "user-3",
            &json!({"org_id": "org-a", "created_at": "2025-12-01"}),
        );
        let condition = SortCondition::Gte(json!("2025-06-01"));
        let result = gsi.query(&json!("org-a"), Some(&condition), false);
        assert_eq!(result, vec!["user-2".to_string(), "user-3".to_string()]);
    }

    #[test]
    fn hash_range_between_condition() {
        let mut gsi = Gsi::new("by_org", "org_id", Some("created_at"));
        gsi.put(
            "user-1",
            &json!({"org_id": "org-a", "created_at": "2025-01-01"}),
        );
        gsi.put(
            "user-2",
            &json!({"org_id": "org-a", "created_at": "2025-06-01"}),
        );
        gsi.put(
            "user-3",
            &json!({"org_id": "org-a", "created_at": "2025-12-01"}),
        );
        let condition = SortCondition::Between(json!("2025-03-01"), json!("2025-09-01"));
        let result = gsi.query(&json!("org-a"), Some(&condition), false);
        assert_eq!(result, vec!["user-2".to_string()]);
    }

    #[test]
    fn update_reindexes_sort_key() {
        let mut gsi = Gsi::new("by_org", "org_id", Some("created_at"));
        gsi.put(
            "user-1",
            &json!({"org_id": "org-a", "created_at": "2025-01-15"}),
        );
        gsi.update(
            "user-1",
            &json!({"org_id": "org-a", "created_at": "2025-01-15"}),
            &json!({"org_id": "org-a", "created_at": "2025-06-01"}),
        );
        let old_condition = SortCondition::Eq(json!("2025-01-15"));
        let new_condition = SortCondition::Eq(json!("2025-06-01"));
        let old_result = gsi.query(&json!("org-a"), Some(&old_condition), false);
        let new_result = gsi.query(&json!("org-a"), Some(&new_condition), false);
        assert!(old_result.is_empty());
        assert_eq!(new_result, vec!["user-1".to_string()]);
    }
}
